### DISCUSSION TITLE
操作 mysql 表时 replace into 或 insert into on duplicate key update 支持

### DISCUSSION THREAD

**Discussion by liuanxin** [apache/doris#21605] [RFC]
当使用 jdbc_catelog 且目标库是 mysql 时, 如果要将 doris 的表数据同步至 mysql, 当遇到唯一索引时, 是否能支持 replace into 或者 insert into uplicate key update 语句? 不然的话, 每次都还得先去 mysql 中删除对应的数据, 然后又不确定该删哪些

**Comment by yagagagaga** [comment]
目前还不支持

**Reply by liuanxin** [reply]
那请问是否有计划支持? 有这个功能会省很多事情, 针对外部库表的 cud 操作在 sql 层面的检查原则上来说由目标库来保证是更为合理的, 现在折中的做法是每次操作前都删掉 mysql 的数据再写入, 这样好繁琐

**Reply by yagagagaga** [reply]
你可以从Doris写到一张临时表，然后再在mysql里使用insert into on duplicate key update到目标表

**Reply by yagagagaga** [reply]
@liuanxin 请问是什么场景需要回写到mysql呢？

**Reply by liuanxin** [reply]
有些历史需求, 最终的统计结果会写到 mysql 表中, 改造这一部分的功能时, 最好的兼容就是基于 doris 做的统计功能弄好之后, 再将结果数据同步到 mysql, 这个场景应该挺常见的吧.

另外还想问一下, 当基于 jdbc_catalog 回写数据到 mysql 时, 可以支持向量化引擎吗? 我见现有的写入还是 doris 组装好数据后构建了单条 insert into 后实现的(在 mysql 中 show full processlist 确认), 后续是否会支持 insert into ... values(...), (...), (...) 这样的批量方式(向量化)呢?

**Reply by yagagagaga** [reply]
由于Doris 是支持mysql协议的，是否可以直接让用户通过Doris来查统计报表，这样就不用回写到mysql了。

回写到mysql的话，在计算的时候是支持向量化的。

**Reply by liuanxin** [reply]
这样需要改造的地方会线性增加, 比如统计表有很多是 bitmap 列, 在同步到 mysql 时会使用 `bitmap_count(field)` 写进 mysql, 如果全部改造, 数据源要调整, 包要调整, sql 也要调整, 现阶段同步是最快的兼容方式.

回写到 mysql 时, 向量化也已经是内置了吗? 还是需要做什么样的配置? 我见在 doris 中(2.0-beta)执行 `insert into jdbc_mysql... select ... from doris_table` 后, 去到 mysql 执行 `show full processlist` 时, 都是单条的 `insert into ... values ()`, 并没有批量的

**Reply by yagagagaga** [reply]
> 这样需要改造的地方会线性增加, 比如统计表有很多是 bitmap 列, 在同步到 mysql 时会使用 `bitmap_count(field)` 写进 mysql, 如果全部改造, 数据源要调整, 包要调整, sql 也要调整, 现阶段同步是最快的兼容方式.
> 
> 回写到 mysql 时, 向量化也已经是内置了吗? 还是需要做什么样的配置? 我见在 doris 中(2.0-beta)执行 `insert into jdbc_mysql... select ... from doris_table` 后, 去到 mysql 执行 `show full processlist` 时, 都是单条的 `insert into ... values ()`, 并没有批量的

计算的时候是有向量化的，写入的时候，还是调用 jdbc 来写入，`insert into ... values ()`PrepareStatement addBatch批量插入的。

**Reply by liuanxin** [reply]
`insert into jdbc_mysql... select ... from doris_table` 这个的写入效率真的很一般, mysql 端大量的 insert into 单记录, 几十万级的数据量写了快 10 分钟, 写完之后的 mysql 内部 `replace into` 在 1 秒左右就处理完了. 是否有计划优化成 `insert into ... values ( ... ) , ( ... )` 这样一条 sql 就可以写多条数据并配合 `PrepareStatement addBatch` 的方式?

**Reply by yagagagaga** [reply]
你可以尝试在 jdbc url中尝试添加一个参数rewriteBatchedStatements=true

**Reply by Junfei1101** [reply]
> 目前还不支持

有支持的计划嘛？
